src/server/frontend_wayland/xdg_decoration_unstable_v1.cpp

```diff
@@ -48,7 +73,7 @@ class XdgDecorationManagerV1 : public wayland::XdgDecorationManagerV1
 
 private:
     void get_toplevel_decoration(wl_resource* id, wl_resource* toplevel) override;
-	std::shared_ptr<std::unordered_set<wl_resource*>> toplevels_with_decorations;
+    ToplevelsWithDecorations toplevels_with_decorations;
```

Comment:
It is unclear that `XdgDecorationManagerV1` is guaranteed to outlive all the references you capture.

It is clearer to hold `ToplevelsWithDecorations` via a shared pointer and capture that by copy.

```suggestion
    std::shared_ptr<ToplevelsWithDecorations> const toplevels_with_decorations{std::make_shared<ToplevelsWithDecorations>()};
```